sql: fix DROP USER to detect default privilege dependencies

spilchen opened the pull request:
Prior to this fix, DROP USER would succeed even when a role had created default privilege rules, leaving orphaned default privilege entries that referenced non-existent roles.

The issue occurred because the dependency detection logic only checked for roles that had privileges granted to them, but missed the case where a role was the creator of default privileges through REVOKE operations. When default privileges are created with REVOKE (e.g., "ALTER DEFAULT PRIVILEGES FOR ROLE user REVOKE EXECUTE ON ROUTINES FROM public"), the grantee list becomes empty, so the original logic didn't detect the dependency.

This fix adds explicit detection of creator dependencies for default privileges, ensuring that roles cannot be dropped if they own default privilege rules.

Fixes #150543.

Release note (bug fix): Fixed DROP USER succeeding when a role owned default privileges, which could leave invalid privilege entries in the system.

Epic: None

yuzefovich reviewed (commented):
Thanks! `pg_regress` changes LGTM, will defer to Foundations for a proper review.

> Reviewed 2 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-:-OX-FcRT3N7seqs7cHAo:bdtdg8s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151472)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/drop_role.go` line 611 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX1NHlK9bsevkQYhVBN:-OX1NHlL2bmNKs3S4Zf4:busn5os) ([raw file](https://github.com/cockroachdb/cockroach/blob/c032212655586a1eb09b8f1fe25ee397b099bacb/pkg/sql/drop_role.go#L611)):*
> ```Go
> 	case privilege.Schemas:
> 		objectType = "schemas"
> 	case privilege.Routines:
> ```

nice catch -- was this missing case also causing us to miss checking some privilege references?
___
*[`pkg/sql/drop_role.go` line 653 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX1OZEnDvH9cml8RvdU:-OX1OZEnDvH9cml8RvdV:baaexcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c032212655586a1eb09b8f1fe25ee397b099bacb/pkg/sql/drop_role.go#L653)):*
> ```Go
> 			if hint == "" {
> 				hint = fmt.Sprintf(
> 					"Role %s has default privileges that revoked access. Use ALTER DEFAULT PRIVILEGES FOR ROLE %s GRANT ALL ON %s TO <role> to restore access.",
> ```

nit: the "Role %s has default privileges that revoked access." part might be more clear if it were passed in with `errors.WithDetail`. then we can make the hint similar to the one we have for the other case on 632.

i also had a question: why is the `grantee` not knowable here? might be worth a comment.
___
*[`pkg/sql/logictest/testdata/logic_test/drop_user` line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX1PC1-2yW5eERJIItB:-OX1PC1-2yW5eERJIItC:b-hivo7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c032212655586a1eb09b8f1fe25ee397b099bacb/pkg/sql/logictest/testdata/logic_test/drop_user#L228)):*
> ```
> # Create default privileges with REVOKE
> statement ok
> ALTER DEFAULT PRIVILEGES FOR ROLE default_priv_user REVOKE EXECUTE ON ROUTINES FROM public
> ```

i'm curious - does this only repro when the default is `REVOKE EXECUTE ON ROUTINES FROM public`, or does it affect other privilege kinds and users?

i'm asking since the default privilege logic has special cases for the EXECUTE privilege for the `public` role, since that's how we implement the behavior that all roles have the ability to execute any function by default.

if the orphaned privilege issues is specific to EXECUTE/public, then i think that means we should be able to clean up the detection logic.


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151472)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/drop_role.go` line 611 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX1NHlK9bsevkQYhVBN:-OX3cisA1xyZZdkvKCUL:b-d12mk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c032212655586a1eb09b8f1fe25ee397b099bacb/pkg/sql/drop_role.go#L611)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nice catch -- was this missing case also causing us to miss checking some privilege references?
</blockquote></details>

No, it was just missing the object type in the hint.
___
*[`pkg/sql/drop_role.go` line 653 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX1OZEnDvH9cml8RvdU:-OX3drXT-hbQ3yOqDa4t:b-rn0eyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c032212655586a1eb09b8f1fe25ee397b099bacb/pkg/sql/drop_role.go#L653)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: the "Role %s has default privileges that revoked access." part might be more clear if it were passed in with `errors.WithDetail`. then we can make the hint similar to the one we have for the other case on 632.

i also had a question: why is the `grantee` not knowable here? might be worth a comment.
</blockquote></details>

Based on your other comment I am going to clean it up. So, I think it's safe to assume the `grantee` is `public`
___
*[`pkg/sql/logictest/testdata/logic_test/drop_user` line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX1PC1-2yW5eERJIItB:-OX3nK_l1jdoLdpjwpuL:bj30sid) ([raw file](https://github.com/cockroachdb/cockroach/blob/c032212655586a1eb09b8f1fe25ee397b099bacb/pkg/sql/logictest/testdata/logic_test/drop_user#L228)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i'm curious - does this only repro when the default is `REVOKE EXECUTE ON ROUTINES FROM public`, or does it affect other privilege kinds and users?

i'm asking since the default privilege logic has special cases for the EXECUTE privilege for the `public` role, since that's how we implement the behavior that all roles have the ability to execute any function by default.

if the orphaned privilege issues is specific to EXECUTE/public, then i think that means we should be able to clean up the detection logic.
</blockquote></details>

It applies to `USAGE` on types too. But still with these two cases I think this can be simplified.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: fix DROP USER to detect default privilege dependencies

Prior to this fix, DROP USER would succeed even when a role had created
default privilege rules, leaving orphaned default privilege entries that
referenced non-existent roles.

The issue occurred because the dependency detection logic only checked for
roles that had privileges granted to them, but missed the case where a role
was the creator of default privileges through REVOKE operations. When default
privileges are created with REVOKE (e.g., "ALTER DEFAULT PRIVILEGES FOR ROLE
user REVOKE EXECUTE ON ROUTINES FROM public"), the grantee list becomes empty,
so the original logic didn't detect the dependency.

This fix adds explicit detection of creator dependencies for default privileges,
ensuring that roles cannot be dropped if they own default privilege rules. This
only applies to EXECUTE on functions and USAGE on types, since those two
privileges are automatically granted for new roles.

Fixes #150543.

Release note (bug fix): Fixed DROP USER succeeding when a role owned default
privileges, which could leave invalid privilege entries in the system.

Epic: None

rafiss reviewed (approved):
nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-:-OX9VPeF8rJPkd6-DDRN:b-j4xmku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen and @yuzefovich)

___
*[`pkg/sql/drop_role.go` line 629 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/151472#-OX9UxIE5nY_5TZlwhWW:-OX9UxIE5nY_5TZlwhWX:b-paqxhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad20f760ff35d74161813685359f82b94dbdb5b/pkg/sql/drop_role.go#L629)):*
> ```Go
> 		grantee username.SQLUsername,
> 		isRevoke bool,
> 		privilegeName string,
> ```

nit: instead of a string, can we use `privilege.Kind` as the parameter type, and only convert to string when formatting?


<!-- Sent from Reviewable.io -->
